patches/api/0449-Add-Lifecycle-Event-system.patch

```diff
@@ -0,0 +1,577 @@
++ */
++@ApiStatus.Experimental
++@ApiStatus.NonExtendable
++public interface Registrar {
++}
+diff --git a/src/main/java/io/papermc/paper/plugin/lifecycle/event/registrar/RegistrarEvent.java b/src/main/java/io/papermc/paper/plugin/lifecycle/event/registrar/RegistrarEvent.java
+new file mode 100644
+index 0000000000000000000000000000000000000000..04a3ea76a82c13abc21066cfe515b44627fc33af
+--- /dev/null
++++ b/src/main/java/io/papermc/paper/plugin/lifecycle/event/registrar/RegistrarEvent.java
+@@ -0,0 +1,60 @@
++package io.papermc.paper.plugin.lifecycle.event.registrar;
++
++import io.papermc.paper.plugin.lifecycle.event.types.LifecycleEvent;
++import org.jetbrains.annotations.ApiStatus;
++import org.jetbrains.annotations.Contract;
++import org.jetbrains.annotations.NotNull;
++
++/**
++ * A lifecycle event that exposes a {@link Registrar} of some kind
++ * to allow management of various things. Look at implementations of
++ * {@link Registrar} for an idea of what uses this event.
++ *
++ * @param <R> registrar type
++ * @see Reloadable
++ */
++@ApiStatus.Experimental
++@ApiStatus.NonExtendable
++public interface RegistrarEvent<R extends Registrar> extends LifecycleEvent {
++
++    /**
++     * Get the registrar related to this event.
++     *
++     * @return the registrar
++     */
++    @Contract(pure = true)
++    @NotNull R registrar();
++
++    /**
++     * A lifecycle event that exposes a {@link Registrar} that is
++     * reloadable.
++     *
++     * @param <R>
++     * @see RegistrarEvent
++     */
++    @ApiStatus.Experimental
++    @ApiStatus.NonExtendable
++    interface Reloadable<R extends Registrar> extends RegistrarEvent<R> {
++
++        /**
++         * Get the cause of this reload.
++         *
++         * @return the cause
++         */
++        @Contract(pure = true)
++        @NotNull Cause cause();
++
++        @ApiStatus.Experimental
++        enum Cause {
++            /**
++             * The initial load of the server.
++             */
++            INITIAL,
++            /**
++             * A reload, triggered via one of the various mechanisms like
++             * the bukkit or minecraft reload commands.
++             */
++            RELOAD
++        }
++    }
++}
+diff --git a/src/main/java/io/papermc/paper/plugin/lifecycle/event/types/LifecycleEvent.java b/src/main/java/io/papermc/paper/plugin/lifecycle/event/types/LifecycleEvent.java
+new file mode 100644
+index 0000000000000000000000000000000000000000..78978aa8a841ceb2fac99cecde9cae375a09a103
+--- /dev/null
++++ b/src/main/java/io/papermc/paper/plugin/lifecycle/event/types/LifecycleEvent.java
+@@ -0,0 +1,15 @@
++package io.papermc.paper.plugin.lifecycle.event.types;
++
++import org.jetbrains.annotations.ApiStatus;
++
++/**
++ * Base type for all Lifecycle Events.
++ * <p>
++ * A Lifecycle Event generally is an event which
++ * fires during times when the older event system is not
++ * available, like during early server initialization.
```

Comment:
Lifecycle events are generally fired where the older event system is not available, like during early server initialization.